### PR TITLE
title: [data] feat: Implement extensible data preprocessor registry

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

Implements an extensible data preprocessor registry system, replacing the previous dictionary-based approach. This new plugin-style system allows users to easily register custom data preprocessor functions using the `@register_preprocessor` decorator. Preprocessors are automatically registered upon module import (e.g., when `veomni.data.multimodal` is imported), eliminating the need to manually modify core files like `preprocess.py` to add new data source handling logic. This addresses the pain point of having to edit submodule code when integrating internal or proprietary data sources.

### Checklist Before Starting

- [x] Search for similar PRs: No similar PRs found for a preprocessor registry system.
- [x] Format the PR title as `[data] feat: Implement extensible data preprocessor registry`

### Test

**Unit Tests:**
Comprehensive unit tests have been added in `tests/data/test_preprocessor_registry.py`. These tests confirm:
-   `@register_preprocessor` decorator correctly registers functions.
-   Built-in preprocessors from `preprocess.py` are automatically registered on import.
-   `get_preprocessor`, `list_preprocessors`, `is_preprocessor_registered`, and `get_all_preprocessors` functions operate as expected.
-   Correct `ValueError` is raised for unregistered preprocessors (e.g., "Unknown dataset name").
-   `UserWarning` is issued for duplicate preprocessor registrations.
-   Preprocessors function correctly when retrieved and executed, including the `conv_preprocess` convenience function.

**Note:** Full pytest suite requires GPU runtime environment. Validated via direct module import.

### API and Usage Example

**1. Define and register a custom preprocessor:**

```python
# In your custom module (e.g., veomni/data/multimodal/my_custom_preprocessors.py)
from veomni.data.multimodal import register_preprocessor

@register_preprocessor("my_custom_vqa_source")
def my_custom_vqa_preprocessor(conversations, **kwargs):
    """
    Preprocessor for a custom VQA data source.
    Converts raw conversation data into VeOmni's standardized format.
    """
    question = conversations[0]["value"].replace("<image>", "").strip()
    answer = conversations[1]["value"]
    return [
        ["user", ("image", None), ("text", question)],
        ["assistant", ("text", answer)]
    ]

# Ensure this module is imported somewhere in your project (e.g., in __init__.py
# or your main training script) to trigger registration.
```

**2. Use the preprocessor in your training script:**

```python
# In your training script
# Importing veomni.data.multimodal will automatically register all built-in
# and custom preprocessors from imported modules.
import veomni.data.multimodal
from veomni.data import build_multisource_dataset
from veomni.data.multimodal.preprocess import conv_preprocess # For convenience

# Option 1: Use the conv_preprocess convenience function
my_conversations = [...] # your raw conversation data
result_conv = conv_preprocess("my_custom_vqa_source", my_conversations)

# Option 2: Get the preprocessor function directly
my_preprocessor_func = veomni.data.multimodal.get_preprocessor("my_custom_vqa_source")
result_direct = my_preprocessor_func(my_conversations)

# When building a dataset using a config, specify the source_name:
# dataset = build_multisource_dataset(config)
```

**3. Reference the custom data source in your config:**

```yaml
data:
  datasets:
    - name: my_vqa_data
      source_name: my_custom_vqa_source # Matches the name registered with @register_preprocessor
      data_path: /path/to/my/data
      weight: 1.0
```

### Design & Code Changes

**Architecture:**

The new architecture centralizes preprocessor management, enabling a modular and extensible system:

```
┌─────────────────────────────────────────────────────────────┐
│ 1. User-defined Preprocessor Modules                        │
│    (e.g., veomni/data/multimodal/preprocess.py,             │
│          my_custom_preprocessors.py)                        │
│    - Contains functions decorated with @register_preprocessor("source_name")
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│ 2. veomni/data/multimodal/preprocessor_registry.py          │
│    - Defines `_PREPROCESSOR_REGISTRY` (global dict)         │
│    - Provides `register_preprocessor` decorator             │
│    - Provides `get_preprocessor`, `list_preprocessors`,     │
│      `is_preprocessor_registered`, `get_all_preprocessors`  │
└─────────────────────────────────────────────────────────────┘
                            ▲
                            │ (Functions in preprocessor_registry.py are exported)
                            │
┌─────────────────────────────────────────────────────────────┐
│ 3. veomni/data/multimodal/__init__.py                       │
│    - Imports `preprocess` module to trigger registration of │
│      all built-in preprocessors.                            │
│    - Exports public API from `preprocessor_registry.py`     │
└─────────────────────────────────────────────────────────────┘
                            │
                            ▼
┌─────────────────────────────────────────────────────────────┐
│ 4. Training Scripts / User Code                             │
│    - Imports `veomni.data.multimodal` (triggers auto-reg.)  │
│    - Uses `get_preprocessor("source_name")` or              │
│      `conv_preprocess("source_name", ...)` to access funcs  │
└─────────────────────────────────────────────────────────────┘
```

**New Files:**

*   `veomni/data/multimodal/preprocessor_registry.py`: Core implementation of the preprocessor registry.
*   `tests/data/test_preprocessor_registry.py`: Comprehensive unit tests for the registry system.
*   `docs/tutorials/preprocessor_registry.md`: Updated comprehensive usage guide and tutorial for the new system.

**Modified Files:**

*   `veomni/data/multimodal/__init__.py`: Exports registry functions and imports `preprocess` to enable auto-registration.
*   `veomni/data/multimodal/preprocess.py`: Refactored to use `@register_preprocessor` for all built-in preprocessors; `DATASETS` dictionary removed; `conv_preprocess` updated.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [x] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs/tutorials/preprocessor_registry.md).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: Comprehensive unit tests added in `tests/data/test_preprocessor_registry.py`.

### CODE REVIEW

**Comment by Coach257** [comment]
Great Design. 
TODOs:
1. refactored veomni/data/multimodal/preprocess.py into the same design

**Comment by Coach257** [comment]
Another question is the difference between data_preprocess and dataset.
The main dataset is multisource_dataset/mapping_dataset/iterable_dataset/byted_dataset/streaming dataset, defining how to get_item.
The data_preprocess is how to preprocess one raw data sample before tokenizer/image_process/etc.
Your dataset_registry is actually a data_preprocess_registry.

**Comment by TimYangst** [comment]
> Another question is the difference between data_preprocess and dataset. The main dataset is multisource_dataset/mapping_dataset/iterable_dataset/byted_dataset/streaming dataset, defining how to get_item. The data_preprocess is how to preprocess one raw data sample before tokenizer/image_process/etc. Your dataset_registry is actually a data_preprocess_registry.

Thanks for the review! Yes, it makes sense, I rename it as preprocessor_registry and update all the related annotation and documents.

File: docs/tutorials/preprocessor_registry.md [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+  datasets:
+    - name: my_data
+      source_name: my_custom_source  # Matches @register_preprocessor name
+      data_path: /path/to/my/dataset
+      weight: 1.0
+```
+
+## Architecture
+
+### Registration Flow
```
Comment (piyifan123) [inline docs/tutorials/preprocessor_registry.md:87]: nit: after conv_preprocess, we also have process_sample and collate function, should we eventually have a doc to explain the whole process? for this doc, maybe just focus on preprocess and register?

File: docs/tutorials/preprocessor_registry.md [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+## Best Practices
+
+1. **Naming Convention**: Use descriptive, lowercase names for preprocessors (e.g., `internal_vqa`, `custom_ocr`).
+2. **Documentation**: Add docstrings to your preprocessor explaining its expected input format and what it does.
+3. **Error Handling**: Add validation for the input format if it's complex, and provide clear error messages.
+4. **Testing**: Write unit tests for your preprocessors.
+5. **Reusability**: Extract common logic into helper functions that are not decorated.
+6. **Multiple Aliases**: Use multiple `@register_preprocessor` decorators if a preprocessor can be used for different but compatible data sources.
+
+## Usage in Training Scripts
```
Comment (piyifan123) [inline docs/tutorials/preprocessor_registry.md:421]: can you include a pointer to an e2e example? i.e., the trainer script, the process function, the preprocess function.

File: veomni/data/multimodal/preprocess.py [hunk 3]
```diff
@@ -315,39 +362,79 @@ def voice_assistant_preprocess(conversations, **kwargs):
...
+        if "<image>" in value:
+            value = value.replace("<image>", "").strip()
+            constructed_conversation.append([role, ("image", None), ("text", value)])
+        else:
+            constructed_conversation.append([role, ("text", value)])
+
+    return constructed_conversation
+
+
+# Add more custom datasets below as needed
```
Comment (piyifan123) [inline veomni/data/multimodal/preprocess.py]: nit: should we ask users to register to register in there own file?

File: veomni/data/multimodal/preprocess.py [hunk 4]
```diff
@@ -315,39 +362,79 @@ def voice_assistant_preprocess(conversations, **kwargs):
...
+# def your_dataset_preprocess(conversations, **kwargs):
+#     ...
+
+
+# ============================================================================
+# Convenience Functions
+# ============================================================================
+
+
+def conv_preprocess(source: str, conversations, **kwargs):
```
Comment (piyifan123) [inline veomni/data/multimodal/preprocess.py]: nit: should we put all convenient function to preprocess_registry.py?  This file can then be renamed to example_preprocess.py?

File: docs/tutorials/preprocessor_registry.md [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+]
+
+# Output: VeOmni's standardized format (strict)
+constructed_conversation = [
+    ["user", ("image", None), ("text", "What is this?")],
+    ["assistant", ("text", "A cat.")]
+]
+```
+
+### Supported Modalities
```
Comment (piyifan123) [inline docs/tutorials/preprocessor_registry.md:133]: instead of replicating the example preprocess here? can we just point users to our example preprocess functions and let them read code directly?

File: tests/data/test_preprocessor_registry.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    import warnings
+
+    from veomni.data.multimodal.preprocessor_registry import register_preprocessor
+
+    @register_preprocessor("test_dup_preprocessor_1")
+    def test_preprocess1(conversations, **kwargs):
+        return [["first"]]
+
+    # Second registration should warn, not error
+    with warnings.catch_warnings(record=True) as w:
```
Comment (piyifan123) [inline tests/data/test_preprocessor_registry.py]: should duplication just throw an error? warning seems not enough?

File: tests/data/test_preprocessor_registry.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+    # Test internal_vqa from preprocess.py
+    test_conversations = [
+        {"from": "human", "value": "<image> What is shown in the image?"},
+        {"from": "gpt", "value": "A detailed answer..."},
+    ]
+    preprocessor = get_preprocessor("internal_vqa")
+    result = preprocessor(test_conversations)
+    assert isinstance(result, list)
+    assert len(result) == 2
```
Comment (piyifan123) [inline tests/data/test_preprocessor_registry.py]: can we just test the execution of the custom processor works properly?

File: veomni/data/multimodal/preprocessor_registry.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        def sharegpt4v_pretrain_preprocess(conversations, **kwargs):
+            # ... preprocessing logic
+            return constructed_conversation
+    """
+
+    def decorator(func: Callable) -> Callable:
+        if name in _PREPROCESSOR_REGISTRY:
+            import warnings
+
+            warnings.warn(f"Preprocessor for '{name}' is already registered. Overwriting with new preprocessor.")
```
Comment (piyifan123) [inline veomni/data/multimodal/preprocessor_registry.py]: can we throw an error here? it does not sounds right to have multiple processor with the same name.

File: docs/tutorials/preprocessor_registry.md (hunk 1 above)
Comment (Coach257) [reply docs/tutorials/preprocessor_registry.md:87]: We need a tutorial/data.md about the pipeline: conv_preprocess -> dataset.transform -> collate_function -> dataloader, each linked to a xxx.md about how to use and how to customize. (Maybe in next pr)

File: veomni/data/multimodal/preprocess.py (hunk 3 above)
Comment (Coach257) [reply veomni/data/multimodal/preprocess.py]: All in one file is acceptable

File: veomni/data/multimodal/preprocess.py (hunk 4 above)
Comment (Coach257) [reply veomni/data/multimodal/preprocess.py]: I think we could take `sharegpt4v` as the example here and delete internal_vqa

File: tests/data/test_preprocessor_registry.py (hunk 7 above)
Comment (Coach257) [reply tests/data/test_preprocessor_registry.py]: I think the whole ci test can be simple as :
1. import registry
2. register a customized preprocess
3. process a sample data with an exit preprocess and the customzied preprocess
4. assert the output follows the interleave structure

File: veomni/data/multimodal/preprocessor_registry.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    List all registered dataset source names.
+
+    Returns:
+        List of dataset source names (sorted)
+    """
+    get_all_preprocessors()
+    return sorted(_PREPROCESSOR_REGISTRY.keys())
+
+
+def is_preprocessor_registered(name: str) -> bool:
```
Comment (Coach257) [inline veomni/data/multimodal/preprocessor_registry.py:142]: I prefer make all the registry functions as a class. Check [chat_template_registry](https://github.com/ByteDance-Seed/VeOmni/blob/c57a165a36939431d5188a82d06b7569d307a704/veomni/data/chat_template.py#L252).
You can keep the decorator.

File: docs/tutorials/preprocessor_registry.md (hunk 1 above)
Comment (TimYangst) [reply docs/tutorials/preprocessor_registry.md:87]: Sounds good, let me focus on the processor register in this PR, and have a delicated PR for the document enhancement.

File: docs/tutorials/preprocessor_registry.md (hunk 2 above)
Comment (TimYangst) [reply docs/tutorials/preprocessor_registry.md:421]: Done

File: docs/tutorials/preprocessor_registry.md (hunk 5 above)
Comment (TimYangst) [reply docs/tutorials/preprocessor_registry.md:133]: Done.

File: veomni/data/multimodal/preprocess.py (hunk 3 above)
Comment (TimYangst) [reply veomni/data/multimodal/preprocess.py]: I think in Veomni all the preprocessors should be registered in a single file. As for other project that depend on Veomni, they could have their own delicated file to register their own custom preprocessor.

File: veomni/data/multimodal/preprocess.py (hunk 4 above)
Comment (TimYangst) [reply veomni/data/multimodal/preprocess.py]: Done, I moved the conv_preprocess to the preprocess_registry.py, and left a placeholder comment for users to register their new preprocessors here.

File: tests/data/test_preprocessor_registry.py (hunk 6 above)
Comment (TimYangst) [reply tests/data/test_preprocessor_registry.py]: Done.

File: tests/data/test_preprocessor_registry.py (hunk 7 above)
Comment (TimYangst) [reply tests/data/test_preprocessor_registry.py]: I added an e2e functional test below, which includes all the processers.

File: veomni/data/multimodal/preprocessor_registry.py (hunk 8 above)
Comment (TimYangst) [reply veomni/data/multimodal/preprocessor_registry.py]: Good point. I think we may need a follow-up PR to clean up the config and the processor mapping. Historically, some datasets shared the same preprocessor in preprocess.py (e.g. "DenseFusion-1M" and "DenseFusion-4V-100k" both use densefusion_preprocess), so multiple names pointing to the same processor wasn’t uncommon.

I’d suggest keeping the current behavior for now, and once we introduce a clearer, explicit mapping between datasource and processor name in the config, we can do a proper cleanup.
